File: syncflow-java/syncflow-bom/src/main/java/com/syncflow/bom/controller/bom/BomController.java
package com.syncflow.bom.controller.bom;

import com.syncflow.bom.dto.*;
import com.syncflow.bom.service.BomService;
import com.syncflow.common.enums.ErrorCode;
import com.syncflow.common.exception.BusinessException;
import com.syncflow.common.result.Result;
import com.syncflow.workflow.entity.ChangeRequest;
import com.syncflow.workflow.service.ChangeRequestService;
import com.syncflow.workflow.service.WorkflowService;
import com.syncflow.common.util.SecurityUtils;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

/**
 * BOM management controller.
 */
@RestController
@RequestMapping("/api/boms")
public class BomController {

    private final BomService bomService;
    private final ChangeRequestService changeRequestService;
    private final WorkflowService workflowService;
    private final ObjectMapper objectMapper;

    public BomController(BomService bomService, ChangeRequestService changeRequestService,
                         WorkflowService workflowService, ObjectMapper objectMapper) {
        this.bomService = bomService;
        this.changeRequestService = changeRequestService;
        this.workflowService = workflowService;
        this.objectMapper = objectMapper;
    }

    /**
     * List BOMs, optionally filtered by project.
     */
    @GetMapping
    public Result<List<BomVO>> listBoms(
            @RequestParam(required = false) Long projectId) {
        List<BomVO> result = bomService.listBoms(projectId);
        return Result.success(result);
    }

    /**
     * BOM detail.
     */
    @GetMapping("/{id}")
    public Result<BomVO> getBomDetail(@PathVariable Long id) {
        BomVO vo = bomService.getBomDetail(id);
        return Result.success(vo);
    }

    /**
     * BOM item tree structure.
     */
    @GetMapping("/{id}/structure")
    public Result<List<BomItemTreeVO>> getBomStructure(@PathVariable Long id) {
        List<BomItemTreeVO> tree = bomService.getBomStructure(id);
        return Result.success(tree);
    }

    /**
     * Create a new BOM.
     */
    @PostMapping
    public Result<BomVO> createBom(@Valid @RequestBody CreateBomDTO dto) {
        BomVO vo = bomService.createBom(dto);
        return Result.success(vo);
    }

    /**
     * Add a BOM item to a BOM.
     */
    @PostMapping("/{id}/items")
    public Result<BomItemTreeVO> addBomItem(@PathVariable Long id,
                                            @Valid @RequestBody CreateBomItemDTO dto) {
        BomItemTreeVO vo = bomService.addBomItem(id, dto);
        return Result.success(vo);
    }

    /**
     * Update a BOM item.
     */
    @PutMapping("/items/{itemId}")
    public Result<BomItemTreeVO> updateBomItem(@PathVariable Long itemId,
                                               @Valid @RequestBody CreateBomItemDTO dto) {
        BomItemTreeVO vo = bomService.updateBomItem(itemId, dto);
        return Result.success(vo);
    }

    /**
     * Delete a BOM item (and its children).
     */
    @DeleteMapping("/items/{itemId}")
    public Result<Void> deleteBomItem(@PathVariable Long itemId) {
        bomService.deleteBomItem(itemId);
        return Result.success();
    }

    /**
     * Submit a BOM for approval.
     */
    @PostMapping("/{id}/submit-approval")
    public Result<Void> submitForApproval(@PathVariable Long id) {
        bomService.submitForApproval(id);
        return Result.success();
    }

    /**
     * Withdraw a pending BOM approval.
     */
    @PostMapping("/{id}/withdraw-approval")
    public Result<Void> withdrawApproval(@PathVariable Long id) {
        bomService.withdrawApproval(id);
        return Result.success();
    }

    /**
     * Save a new version of the BOM.
     */
    @PostMapping("/{id}/save-version")
    public Result<BomVO> saveVersion(@PathVariable Long id,
                                     @RequestParam(required = false) String changeSummary) {
        BomVO vo = bomService.saveVersion(id, changeSummary);
        return Result.success(vo);
    }

    /**
     * Get version history of a BOM.
     */
    @GetMapping("/{id}/versions")
    public Result<List<BomVersionVO>> getVersionHistory(@PathVariable Long id) {
        List<BomVersionVO> versions = bomService.getVersionHistory(id);
        return Result.success(versions);
    }

    /**
     * Compare two versions of a BOM and return item-level diffs.
     * <p>
     * Example: GET /api/boms/42/compare?v1=1.0&v2=1.1
     */
    @GetMapping("/{id}/compare")
    public Result<BomVersionCompareVO> compareVersions(
            @PathVariable Long id,
            @RequestParam String v1,
            @RequestParam String v2) {
        return Result.success(bomService.compareVersions(id, v1, v2));
    }

    /**
     * Roll back a BOM to a previously saved version.
     * Example: POST /api/boms/42/rollback?targetVersion=1.0
     */
    @PostMapping("/{id}/rollback")
    public Result<Void> rollbackVersion(@PathVariable Long id,
                                        @RequestParam String targetVersion) {
        bomService.rollbackVersion(id, targetVersion);
        return Result.success();
    }

    /**
     * Get change requests for a BOM.
     */
    @GetMapping("/{id}/change-requests")
    public Result<List<ChangeRequest>> getChangeRequests(@PathVariable Long id) {
        List<ChangeRequest> requests = changeRequestService.getRequestsByObject("BOM_CHANGE", id);
        return Result.success(requests);
    }

    /**
     * Submit a BOM change request and start the CHANGE_APPROVAL workflow.
     * changeData is a JSON string describing the change (ADD_ITEM / UPDATE_ITEM / DELETE_ITEM).
     */
    @PostMapping("/{id}/change-requests")
    public Result<Map<String, Object>> createChangeRequest(
            @PathVariable Long id,
            @RequestBody Map<String, Object> body) {
        String changeType = (String) body.get("changeType");
        if (changeType == null) {
            throw new BusinessException(ErrorCode.PARAM_ERROR, "changeType is required");
        }

        // Build changeData JSON from the request body fields
        String changeData;
        try {
            Map<String, Object> dataMap = new java.util.LinkedHashMap<>();
            switch (changeType) {
                case "ADD_ITEM" -> {
                    dataMap.put("name", body.get("name"));
                    dataMap.put("sourceType", body.getOrDefault("sourceType", "MADE"));
                    dataMap.put("quantity", body.getOrDefault("quantity", 1));
                    if (body.get("specification") != null) dataMap.put("specification", body.get("specification"));
                    if (body.get("materialCode") != null) dataMap.put("materialCode", body.get("materialCode"));
                    if (body.get("unitOfMeasure") != null) dataMap.put("unit", body.get("unitOfMeasure"));
                    if (body.get("parentId") != null) dataMap.put("parentId", body.get("parentId"));
                }
                case "UPDATE_ITEM" -> {
                    dataMap.put("itemId", body.get("itemId"));
                    if (body.get("name") != null) dataMap.put("name", body.get("name"));
                    if (body.get("sourceType") != null) dataMap.put("sourceType", body.get("sourceType"));
                    if (body.get("quantity") != null) dataMap.put("quantity", body.get("quantity"));
                    if (body.get("specification") != null) dataMap.put("specification", body.get("specification"));
                    if (body.get("materialCode") != null) dataMap.put("materialCode", body.get("materialCode"));
                }
                case "DELETE_ITEM" -> dataMap.put("itemId", body.get("itemId"));
                default -> throw new BusinessException(ErrorCode.PARAM_ERROR, "Unknown changeType: " + changeType);
            }
            changeData = objectMapper.writeValueAsString(dataMap);
        } catch (BusinessException e) {
            throw e;
        } catch (Exception e) {
            throw new BusinessException(ErrorCode.PARAM_ERROR, "Failed to serialize changeData: " + e.getMessage());
        }

        String changeSummary = (String) body.getOrDefault("description", changeType);
        Long requestedBy = SecurityUtils.getUserId();

        Long crId = changeRequestService.createRequest("BOM_CHANGE", id, changeType,
                changeData, changeSummary, requestedBy);

        BomVO bom = bomService.getBomDetail(id);
        workflowService.startProcess("CHANGE_APPROVAL", crId, "BOM_CHANGE",
                changeSummary, bom.getProjectId(), requestedBy);

        return Result.success(Map.of("changeRequestId", crId));
    }
}

File: v1-nestjs/server/src/bom/bom.controller.ts
import {
  Controller,
  Get,
  Post,
  Patch,
  Delete,
  Param,
  Body,
  Query,
} from '@nestjs/common';
import { BomService } from './bom.service';
import { CreateBomItemDto } from './dto/create-bom-item.dto';
import { UpdateBomItemDto } from './dto/update-bom-item.dto';

@Controller('bom')
export class BomController {
  constructor(private readonly bomService: BomService) {}

  @Get()
  async findAll(@Query('projectId') projectId: string) {
    const items = await this.bomService.findAll(projectId);
    return { code: 0, data: items };
  }

  @Get('tree')
  async findTree(@Query('projectId') projectId: string) {
    const tree = await this.bomService.findTree(projectId);
    return { code: 0, data: tree };
  }

  @Get('versions')
  async getVersions(@Query('projectId') projectId: string) {
    const versions = await this.bomService.getVersions(projectId);
    return { code: 0, data: versions };
  }

  @Get('compare')
  async compareVersions(
    @Query('projectId') projectId: string,
    @Query('v1') v1: string,
    @Query('v2') v2: string,
  ) {
    const result = await this.bomService.compareVersions(
      projectId,
      parseInt(v1, 10),
      parseInt(v2, 10),
    );
    return { code: 0, data: result };
  }

  @Get('export')
  async exportToData(@Query('projectId') projectId: string) {
    const items = await this.bomService.exportToData(projectId);
    return { code: 0, data: items };
  }

  @Get(':id')
  async findOne(@Param('id') id: string) {
    const item = await this.bomService.findOne(id);
    return { code: 0, data: item };
  }

  @Post()
  async create(@Body() body: CreateBomItemDto) {
    const item = await this.bomService.create(body);
    return { code: 0, data: item };
  }

  @Post('import')
  async importFromData(@Body() body: { projectId: string; items: Array<{ name: string; partNumber: string; specification?: string; supplier?: string; unit?: string; unitPrice?: number; quantity?: number }> }) {
    const result = await this.bomService.importFromData(body.projectId, body.items);
    return { code: 0, data: result };
  }

  @Post('versions')
  async createVersion(
    @Body() body: { projectId: string; description?: string },
  ) {
    const version = await this.bomService.createVersion(
      body.projectId,
      body.description,
    );
    return { code: 0, data: version };
  }

  @Post('rollback')
  async rollbackVersion(
    @Body() body: { projectId: string; targetVersion: number },
  ) {
    const version = await this.bomService.rollbackVersion(
      body.projectId,
      body.targetVersion,
    );
    return { code: 0, data: version };
  }

  @Patch(':id')
  async update(@Param('id') id: string, @Body() body: UpdateBomItemDto) {
    const item = await this.bomService.update(id, body);
    return { code: 0, data: item };
  }

  @Delete(':id')
  async remove(@Param('id') id: string) {
    const item = await this.bomService.remove(id);
    return { code: 0, data: item };
  }
}
